### ISSUE TITLE
Memory Efficient FP 16 Training

### ISSUE THREAD

**Issue by rajgar114** [huggingface/transformers#12084]
# 🚀 Feature request

Fairseq uses memory efficient FP 16 training as explained in https://arxiv.org/pdf/1904.01038.pdf.

## Motivation
Generally the model requires high end GPU's to fine-tune on larger length datasets. Using memory efficient FP 16 we can reduce the need of high GPU's and thus models can be fine-tune without OOM problems.

**Comment by LysandreJik** [comment]
Is "memory efficient" fp16 different to the fp16 available in the training scripts?

**Comment by rajgar114** [comment]
Yes, both are different.
https://github.com/pytorch/fairseq/issues/2907

**Comment by LysandreJik** [comment]
Thanks for the link! cc @sgugger @stas00

**Comment by stas00** [comment]
@rajgar114, if you think you have found some optimization that we can improve performance with I'm all ears, but I don't think this is it.

This is a paper from 2019. We are in 2021.

**Comment by stas00** [comment]
Not sure how to evaluate the effectiveness this proposal, it would be helpful to have some case studies that show the actual improvements.

I asked around and someone reported that someone mentioned this was useful for certain models, but it'd help to know which and how they were trained so that there is an actual proven setup to work with. 

I found pytext included a variation of it here: https://github.com/facebookresearch/pytext/commit/c6d13acbafc856fdc0291bf6608d6f318b6690d2, but I can't find any other references via google, which is not very encouraging. But we don't know whether other implementations use the same name.

But also let's ask this, @rajgar114, would you like to try to work on this and submit a PR when you have something working?

**Comment by rajgar114** [comment]
`I think you must be referring to this comment pytorch/fairseq#2907 (comment) ?`

Yes, @stas00 I was referring to this comment [pytorch/fairseq#2907](https://github.com/pytorch/fairseq/issues/2907#issuecomment-729722676) only.

`Appears to be added 2 years ago.`

No doubt the paper is quite old but I have found some instances that memory efficient fp16 training worked for people having low end GPU's. Here is an example:
https://bleepcoder.com/fairseq/551167214/oom-while-trying-to-train-bart 

`Do you think it's actually something that's worth investigating?`

I am not completely sure that how much impact it can cause on the reduction in GPU memory consumption. We should definitely ask some experienced guys and also try to compare and analyze the results with & without using memory efficient fp16 version. 

`Would you like to try to work on this and submit a PR when you have something working?`

@stas00 Thanks for giving this wonderful opportunity. I would love to work on open source projects. But I can't because of my busy schedule, I would not be able to spend time on this project. Sorry for that.

**Comment by stas00** [comment]
Thank you for the feedback, @rajgar114 and letting us know that you'd love to work on that but reality won't allow that at the moment.

I have added it to https://github.com/huggingface/transformers/issues/12126 so it won't get lost and hopefully it'd find a champion.

From your part what would help is to find specific models/setups where it has been found to be converging well despite the limitations it imposed on itself. So that whoever works on this will have a good chance of succeeding. Thank you.

**Comment by ksopyla** [comment]
I do not know exactly if my case fits this issue, but I have trained seq2seq model with fairseq (v0.10.1) and current transformer 4.9. In both cases I have used encoder-decoder and BART architecture and saw that Fariseq allows me to use a greater batch size 120- 160 (sentences) around 5000 tokens. Transformers library with deepspeed integration handles max 48(sentences).
I have used 3x Geforce 3090 with 24 GB ram. In both cases model size ~70M parameters (6 layers for encoder and decoder, hidden_size=512, ff=2048)
Concludes, fairseq training is almost 3 times faster (measured by the number of tokens seen during training in a fixed time budget).
